File: yong-liang-ge-zhan-shi-xian-dui-lie-lcof.go
type Stack struct {
    Values []int
}

func (this *Stack) Push(value int) {
    this.Values = append(this.Values, value)
}

func (this *Stack) Pop() (int, bool) {
    if len(this.Values) == 0 {
        return -1, false
    }
    value := this.Values[len(this.Values)-1]
    this.Values = this.Values[:len(this.Values)-1]
    return value, true
}

type CQueue struct {
    SIn *Stack
    SOut *Stack
}


func Constructor() CQueue {
    return CQueue{
        SIn: &Stack{[]int{}},
        SOut: &Stack{[]int{}},
    }
}


func (this *CQueue) AppendTail(value int)  {
    this.SIn.Push(value)
}


func (this *CQueue) DeleteHead() int {
    value, ok := this.SOut.Pop()
    if ok {
        return value
    }
    for value, ok := this.SIn.Pop(); ok; value, ok = this.SIn.Pop(){
        this.SOut.Push(value)
    }
    value, _ = this.SOut.Pop()
    return value
}


/**
 * Your CQueue object will be instantiated and called as such:
 * obj := Constructor();
 * obj.AppendTail(value);
 * param_2 := obj.DeleteHead();
 */

File: yong-liang-ge-zhan-shi-xian-dui-lie-lcof.c
typedef struct {
    int* inStack;
    int* ouStack;
    int inStackLen;
    int ouStackLen;
} CQueue;


CQueue* cQueueCreate() {
    CQueue* cq = malloc(sizeof(CQueue));
    cq->inStack = malloc(sizeof(int)*10000);
    cq->ouStack = malloc(sizeof(int)*10000);
    cq->inStackLen = 0;
    cq->ouStackLen = 0;
    return cq;
}

void cQueueAppendTail(CQueue* obj, int value) {
    obj->inStack[obj->inStackLen++] = value;
}

int cQueueDeleteHead(CQueue* obj) {
    if (obj->ouStackLen>0) {
        return obj->ouStack[--obj->ouStackLen];
    }
    while (obj->inStackLen>0) {
        obj->ouStack[obj->ouStackLen++] = obj->inStack[--obj->inStackLen];
    }
    if (obj->ouStackLen>0) {
        return obj->ouStack[--obj->ouStackLen];
    }
    return -1;
}

void cQueueFree(CQueue* obj) {
}

/**
 * Your CQueue struct will be instantiated and called as such:
 * CQueue* obj = cQueueCreate();
 * cQueueAppendTail(obj, value);
 
 * int param_2 = cQueueDeleteHead(obj);
 
 * cQueueFree(obj);
*/
